src/rust/engine/process_execution/src/lib.rs

```diff
@@ -275,6 +276,51 @@ impl InputDigests {
     })
   }
 
+  pub async fn new_from_merged(store: &Store, from: Vec<InputDigests>) -> Result<Self, String> {
+    Ok(Self {
+      complete: store
+        .merge(
+          from
+            .iter()
+            .map(|input_digests| input_digests.complete.clone())
+            .collect(),
+        )
+        .await?,
+      nailgun: store
+        .merge(
+          from
+            .iter()
+            .map(|input_digests| input_digests.nailgun.clone())
+            .collect(),
+        )
+        .await?,
+      input_files: store
+        .merge(
+          from
+            .iter()
+            .map(|input_digests| input_digests.input_files.clone())
+            .collect(),
+        )
+        .await?,
+      immutable_inputs: Itertools::concat(
+        from
+          .iter()
+          .map(|input_digests| input_digests.immutable_inputs.clone()),
+      )
+      .into_iter()
+      .unique()
```

Comment:
This is a `BTreeMap`, so the unique isn't necessary I don't think.

But you _do_ potentially need to validate that they don't have different contents for some keys...?

Reply:
Done, I used an after-the-fact size comparison to make the check speedy (as opposed to looping through and appending one-by-one. I assume `.append` is smarter than looping). We lose the context of _which_ key is duplicate though.